JsonLedgerApiSample/src/main/java/com/example/store/IntegrationStore.java

```diff
@@ -24,6 +28,170 @@ public class IntegrationStore {
+         */
+        private int ingestEvent(Event event0) {
+            if (event0.getActualInstance() instanceof EventOneOf event) {
+                throw new UnsupportedOperationException("Did not expect archive event as we are subscribing using LEDGER_EFFECTS: " + event.toJson());
+            } else if (event0.getActualInstance() instanceof EventOneOf1 event) {
+                return ingestCreateEvent(event.getCreatedEvent());
+            } else if (event0.getActualInstance() instanceof EventOneOf2 event) {
+                return ingestExerciseEvent(event.getExercisedEvent());
+            } else {
+                throw new UnsupportedOperationException("Failed to handle event: " + event0.toJson());
+            }
+        }
+
+        /**
+         * Parse a create event, return its node ID.
+         */
+        private int ingestCreateEvent(CreatedEvent createdEvent) {
+            List<JsInterfaceView> interfaceViews = createdEvent.getInterfaceViews();
+            if (interfaceViews != null) {
+                for (JsInterfaceView view : interfaceViews) {
+                    if (TemplateId.HOLDING_INTERFACE_ID.matchesModuleAndTypeName(view.getInterfaceId())) {
+                        String viewJson = com.example.client.ledger.invoker.JSON.getGson().toJson(view.getViewValue());
+                        HoldingView holding = ConversionHelpers.convertFromJson(viewJson, HoldingView::fromJson);
+                        ingestCreateHoldingEvent(createdEvent.getContractId(), holding, transferInfo);
+                        return createdEvent.getNodeId();
+                    }
+                }
+            }
+            log.finer(() -> "Ignoring create event as it does not implement Holding interface: " + createdEvent.toJson());
+            return createdEvent.getNodeId();
+        }
+
+        /**
+         * Parse an exercise event, return the node ID of the last descendant node.
+         */
+        private int ingestExerciseEvent(ExercisedEvent exercisedEvent) {
+            // Parse effect of the exercise node itself
+            if (!exercisedEvent.getConsuming()) {
+                log.finer(() -> "Ignoring non-consuming exercise event: " + exercisedEvent.toJson());
+            } else {
+                String cid = exercisedEvent.getContractId();
+                HoldingView oldView = activeHoldings.remove(cid);
+                if (oldView == null) {
+                    log.fine(() -> "Ignoring consuming exercise event for untracked contract: " + exercisedEvent.toJson());
+                } else {
+                    log.info("Processing consuming choice " + exercisedEvent.getChoice() + " on tracked holding " + cid);
+                    if (transferInfo != null && !transferInfo.sender.equals(treasuryParty)) {
+                        // This would only happen if there are multiple intermediate holdings owned by the treasury
```

Comment:
I was not able to validate the correlation between the if statement and the conclusion in the comment, perhaps I don't understand well enough the context in which this if statement executes.

I would have thought that this condition would simply mean "sender is making a deposit", no?

Reply:
adjusted the comment to:

> // TODO: consider just forbidding this case, and ignoring a deposit that triggers it
>
> // Seeing a consuming exercise on a tracked holding (i.e., one owned by the treasury party)
> // for an incoming transfer is surprising. Most token admins will not use such a pattern,
> // but some might have workflows where the holding is both
> // created and archived within the same transfer.
> //
> // The safe option is to process it as a debit so that create + archive pairs net to zero.

I expect we'll have to work some more of this once we have more test data together.
I have some ideas to track a transfer log to make this kind of special case more explicit.